### PR TITLE
Fix inaccurate navigation

### PR DESCRIPTION
**Pull request by ahmedriad1**
### Summary
Navigation in the translation view was totally inaccurate. This is due to:
1. We render virtuoso list items as pages, not verses, to fetch data in one place. Therefore, in `useScrollToVirtualizedTranslationView`, when we scroll to an ayah, we try to approximate where it is. 
2. When ayahs are loaded, the scroll position changes due to a difference in the real size vs. the skeleton one, therefore, showing the user a different ayah. 

### Solution
1. We render ayahs individually, and SWR will cache requests for the same page. Now, when we navigate, we can pass the ayah index directly to Virtuoso.
2. When we scroll to un-loaded ayahs, we re-adjust the scroll position after they're loaded.

### CODE REVIEW

File: src/components/QuranReader/TranslationView/TranslationViewVerse/hooks/useDedupedFetchVerse.ts [hunk 1]
```diff
@@ -1,60 +1,55 @@
...
+  mushafId: Mushaf;
+  selectedTranslations: number[];
+  setApiPageToVersesMap: (data: Record<number, Verse[]>) => void;
+  verseIdx: number;
 }
 
-const TranslationPage: React.FC<Props> = ({
-  totalPages,
-  pageNumber,
+const useDedupedFetchVerse = ({
```
Comment (osamasayed) [inline src/components/QuranReader/TranslationView/TranslationViewVerse/hooks/useDedupedFetchVerse.ts:37]: Add a comment

File: src/components/QuranReader/TranslationView/TranslationViewVerse/hooks/useDedupedFetchVerse.ts [hunk 2]
```diff
@@ -94,21 +94,10 @@ const TranslationPage: React.FC<Props> = ({
     verseFetcher,
     {
       fallbackData: shouldUseInitialData ? initialData.verses : undefined,
-      revalidateOnMount: !shouldUseInitialData,
+      // revalidateOnMount: !shouldUseInitialData,
```
Comment (osamasayed) [inline src/components/QuranReader/TranslationView/TranslationViewVerse/hooks/useDedupedFetchVerse.ts:97]: Remove this line

File: src/components/QuranReader/TranslationView/hooks/useScrollToVirtualizedVerse.ts [hunk 3]
```diff
@@ -36,34 +15,90 @@ const getVersePositionWithinAPage = (
...
   const router = useRouter();
+  const [shouldReadjustScroll, setShouldReadjustScroll] = useState(false);
+
   const { startingVerse } = router.query;
+  const startingVerseNumber = Number(startingVerse);
+  // if the startingVerse is a valid integer and is above 1
+  const isValidStartingVerse =
+    startingVerseNumber && Number.isInteger(startingVerseNumber) && startingVerseNumber > 0;
+
+  const scrollToArabicVerse = useCallback(
```
Comment (osamasayed) [inline src/components/QuranReader/TranslationView/hooks/useScrollToVirtualizedVerse.ts:36]: rename to `scrollToVerseCellStart`

File: src/components/QuranReader/TranslationView/hooks/useScrollToVirtualizedVerse.ts [hunk 4]
```diff
@@ -36,34 +15,90 @@ const getVersePositionWithinAPage = (
...
+  apiPageToVersesMap: Record<number, Verse[]>,
   versesPerPage: number,
 ) => {
   const router = useRouter();
+  const [shouldReadjustScroll, setShouldReadjustScroll] = useState(false);
+
   const { startingVerse } = router.query;
+  const startingVerseNumber = Number(startingVerse);
+  // if the startingVerse is a valid integer and is above 1
+  const isValidStartingVerse =
```
Comment (osamasayed) [inline src/components/QuranReader/TranslationView/hooks/useScrollToVirtualizedVerse.ts:33]: Re-use `isValidVerseId`

File: src/components/QuranReader/TranslationView/hooks/useScrollToVirtualizedVerse.ts [hunk 5]
```diff
@@ -36,34 +15,90 @@ const getVersePositionWithinAPage = (
...
+    startingVerseNumber && Number.isInteger(startingVerseNumber) && startingVerseNumber > 0;
+
+  const scrollToArabicVerse = useCallback(
+    (verseNumber: number, done?: () => void) => {
+      const verseIndex = verseNumber - 1;
+      virtuosoRef.current.scrollToIndex({
+        index: verseIndex,
+        align: ScrollAlign.Start,
+        offset: -70,
```
Comment (osamasayed) [inline src/components/QuranReader/TranslationView/hooks/useScrollToVirtualizedVerse.ts:42]: Add a comment explaining what 70 stands for

File: src/components/QuranReader/TranslationView/hooks/useScrollToVirtualizedVerse.ts [hunk 6]
```diff
@@ -36,34 +15,90 @@ const getVersePositionWithinAPage = (
...
+      });
+
+      done?.();
```
Comment (osamasayed) [inline src/components/QuranReader/TranslationView/hooks/useScrollToVirtualizedVerse.ts:45]: Remove this

File: src/components/QuranReader/TranslationView/hooks/useScrollToVirtualizedVerse.ts [hunk 7]
```diff
@@ -36,34 +15,90 @@ const getVersePositionWithinAPage = (
...
+    },
+    [virtuosoRef],
+  );
 
   useEffect(() => {
```
Comment (osamasayed) [inline src/components/QuranReader/TranslationView/hooks/useScrollToVirtualizedVerse.ts:50]: Add a comment

File: src/components/QuranReader/TranslationView/hooks/useScrollToVirtualizedVerse.ts [hunk 8]
```diff
@@ -36,34 +15,90 @@ const getVersePositionWithinAPage = (
...
+    }
+  }, [quranReaderDataType, startingVerseNumber, isValidStartingVerse, scrollToArabicVerse]);
+
+  const oldApiPageToVersesMap = useRef<Record<number, Verse[]>>(apiPageToVersesMap);
+
+  // this effect handles the case when the user navigates to a verse that is not yet loaded
+  // we need to wait for the verse to be loaded and then scroll to it
+  useEffect(() => {
+    if (quranReaderDataType === QuranReaderDataType.Chapter && isValidStartingVerse) {
+      const pageNumber = Math.ceil((startingVerseNumber + 1) / versesPerPage);
```
Comment (osamasayed) [inline src/components/QuranReader/TranslationView/hooks/useScrollToVirtualizedVerse.ts:65]: This should be moved to a common util with `useDedepuedFetchVerse.pageNumber`

File: src/components/QuranReader/TranslationView/hooks/useScrollToVirtualizedVerse.ts [hunk 9]
```diff
@@ -36,34 +15,90 @@ const getVersePositionWithinAPage = (
...
+
+      const isFirstVerseInPage = (startingVerseNumber + 1) % versesPerPage === 1;
+      const isLastVerseInPage = (startingVerseNumber + 1) % versesPerPage === 0;
+
+      const isNewPageLoaded = (page: number) => {
+        return !oldApiPageToVersesMap.current[page] && apiPageToVersesMap[page];
+      };
+
+      if (shouldReadjustScroll) {
+        if (
```
Comment (osamasayed) [inline src/components/QuranReader/TranslationView/hooks/useScrollToVirtualizedVerse.ts:75]: Add a comment

File: src/components/QuranReader/TranslationView/hooks/useScrollToVirtualizedVerse.ts [hunk 10]
```diff
@@ -36,34 +15,90 @@ const getVersePositionWithinAPage = (
...
+          isNewPageLoaded(pageNumber) ||
+          (pageNumber > 1 && isFirstVerseInPage && isNewPageLoaded(pageNumber - 1)) ||
+          (isLastVerseInPage && isNewPageLoaded(pageNumber + 1))
+        ) {
+          scrollToArabicVerse(startingVerseNumber);
+        } else {
+          setTimeout(() => {
+            scrollToArabicVerse(startingVerseNumber);
```
Comment (osamasayed) [inline src/components/QuranReader/TranslationView/hooks/useScrollToVirtualizedVerse.ts:83]: Test if using `useBrowserLayoutEffect` make us not need the timeout

Comment (ahmedriad1) [reply src/components/QuranReader/TranslationView/hooks/useScrollToVirtualizedVerse.ts:83]: unfortunately, it does not :(

**Review by osamasayed — APPROVED** [review]

**Review by osamasayed — APPROVED** [review]

**Review by osamasayed — APPROVED** [review]

**Pull request merged**